cpp/ql/src/semmle/code/cpp/controlflow/BasicBlocks.qll

```diff
@@ -245,8 +245,14 @@ class BasicBlock extends ControlFlowNodeBase {
   predicate isReachable() {
     exists(Function f | f.getBlock() = this)
     or
-    exists(TryStmt t, BasicBlock tryblock | this = t.getACatchClause() and tryblock.isReachable() and tryblock.contains(t))
-    or
+    exists(TryStmt t, BasicBlock tryblock |
+      (
+        this = t.getACatchClause() or
+        this.(Handler).getTryStmt() = t
```

Comment:
Why is `this = t.getACatchClause() or` necessary? Aren't all `CatchBlock`s reachable from their handler through the BB successor graph?

Reply:
It was necessary to get correct result, I suspect what may be happening is that not all `Handler`s and/or `CatchBlock`s are `BasicBlock`s.

Reply:
I wrote a quick query to find out whether all `Handlers` are `BasicBlock`s. LibreOffice has 453 such blocks, and Wireshark has 65.

It appears that all `CatchBlock`s have a `Handler` predecessor -- the following query has no results:

```
from CatchBlock catch
where not exists(Handler h | h.getASuccessor() = catch)
select catch
```

We could change `isReachable` to say that if a `BasicBlock` _contains_ a `Handler` of a reachable `try` block then the whole block is reachable, or we could change the construction of primitive basic blocks to say that a `Handler` always starts a new basic block.

What do you think?

Reply:
I don't quite follow your first suggestion.
```
{ // Block A
  try
  { // Block B
  } catch (Exception e) { // Handler C -> CatchBlock D
  }
}
```
The previous code said that if `A` is reachable, `D` is reachable.  My current implementation says that if `A` is reachable, `C` and `D` are reachable.  What are you suggesting?

Your second suggestion is more along the lines I was thinking.  I believe a `CatchBlock` always comes after a `Handler`, but the `Handler` (`C` above) is not necessarily a `BasicBlock`, so we'd have to fix that.  Perhaps by adding a case directly to `primitive_basic_block_entry_node`?

Reply:
My first suggestion was to say that if `A` is reachable then the basic block containing `C` is reachable. This block will most often be `C` itself, but sometimes it will be a block that contains unrelated nodes before `C`, and those might be marked reachable even though they aren't. Because `D` follows `C`, `D` will be marked reachable as long as we mark `C` reachable. This suggestion doesn't do "the right thing", so I'm only suggesting it because it would be a very local and non-disruptive fix.

I think we agree on what's needed for the second suggestion. It's the option that I like the best, as long as there's a pair of comments in `isReachable` and `primitive_basic_block_entry_node` that refer to each other.

Reply:
I've made the change.